### DISCUSSION TITLE
Add SwA / MTOM attachements middleware

### DISCUSSION THREAD

**Discussion by veewee** [phpro/soap-client#357] [Ideas]
Make it possible to send / receive attachements in the soap-client.

- SwA: SOAP Messages with Attachments [Spec](https://www.w3.org/TR/SOAP-attachments)
- MTOM: SOAP Message Transmission Optimization Mechanism [Spec](https://www.w3.org/TR/soap12-mtom/)
  - XOP :  XML-binary Optimized Packaging [Spec](https://www.w3.org/TR/xop10/) 

A sample Jave implementation:
- http://www.codejava.net/java-ee/web-services/using-mtom-to-optimize-binary-data-transfer-with-jax-ws-web-services

Sample PHP implementations:
- https://github.com/debuss/MTOMSoapClient/blob/master/MTOMSoapClient.php
- https://github.com/BeSimple/BeSimpleSoap/blob/master/src/BeSimple/SoapClient/README.md

SOAP-UI docs

- https://www.soapui.org/docs/soap-and-wsdl/attachments/

**Comment by ilgala** [comment]
+1 for veewee... Any workarounds available atm in order to send attachment with MTOM?

**Comment by veewee** [comment]
Hello @IlGala,

Currently there isn't a way nor Middleware for this package. I was thinking about creating one that accepts a filebag for sending / storing files and zend-mime to assemble / extract the message.

For the moment I do not have a use-case for this feature nor the time to work it out. Feel free to implement a middleware solution at any time.

There is also another soap library with a slightly different approach. It does support a subset of SOAP Attachments. You might want to take a look at:
https://github.com/BeSimple/BeSimpleSoap/blob/master/src/BeSimple/SoapClient/README.md

**Comment by veewee** [comment]
If somebody knows of a SOAP service which accepts or provides attachement: that would be a good starting point! :)

**Reply by veewee** [reply]
It turns out 'mockmotor' provides a mock service:

- https://demo.mockmotor.com/console/responses.html?envid=888a35cd-52f0-43eb-a50a-875153831cec&srvid=869c80e7-7b25-4bbe-9184-55d9ccb20df8

Blogs:

- https://mockmotor.com/blog/soap-with-attachments-mock/

```
HTTP/1.1 200 OK
Date: Fri, 01 Jun 2018 02:09:53 GMT
Content-Type: multipart/related; type="text/xml"; start="d29fa4f6db1d4ed08c5b6d39eb7422f4"; boundary="c8f00991f64c479ea05d17b3324e8f21"
Transfer-Encoding: chunked

--c8f00991f64c479ea05d17b3324e8f21
Content-Type: text/xml; charset=UTF-8
Content-ID: d29fa4f6db1d4ed08c5b6d39eb7422f4
X-MockMotor-Delay: 500

<soapenv:Envelope xmlns:urn="urn:GetContractSchema"
                  xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/">
   <soapenv:Header/>
   <soapenv:Body>
      <urn:GetContractResponse>
         <ContractAttachmentRef>Content-ID-1</ContractAttachmentRef>
      </urn:GetContractResponse>
   </soapenv:Body>
</soapenv:Envelope>

--c8f00991f64c479ea05d17b3324e8f21
Content-Type: application/pdf
Content-ID: <Content-ID-1>
Content-Disposition: attachment; name="contract.pdf"; filename="contract.pdf"
Content-Transfer-Encoding: binary

%PDF-1.7
1308 0 obj
<</Linearized 1/L 643967/O 1314/E 57853/N 5/T 643378/H [ 753 544]>>
endobj
...
```

- https://mockmotor.com/blog/soap-with-attachments-mtom-xop-mock/


```
HTTP/1.1 200 OK
Date: Thu, 31 May 2018 03:29:38 GMT
Content-Type: multipart/related; type="application/xop+xml"; boundary="1cfbcfcf5bd24ac4ac8c84e2d431bac7"; start="daf16aee1529464db80c1874f81d01b3"; start-info="text/xml"

--1cfbcfcf5bd24ac4ac8c84e2d431bac7
Content-Type: application/xop+xml; charset=UTF-8; type=text/xml
Content-ID: daf16aee1529464db80c1874f81d01b3

    ... SOAP response payload is here ...
    <data>
        <xop:Include href="cid:ATT1" xmlns:xop="http://www.w3.org/2004/08/xop/include"/>
    </data>
    ...

--1cfbcfcf5bd24ac4ac8c84e2d431bac7
Content-Type: application/pdf
Content-ID: <ATT1>
Content-Disposition: attachment; name="contract.pdf"; filename="contract.pdf"
Content-Transfer-Encoding: binary

%PDF-1.7
1308 0 obj
<</Linearized 1/L 643967/O 1314/E 57853/N 5/T 643378/H [ 753 544]>>
endobj
...
--1cfbcfcf5bd24ac4ac8c84e2d431bac7--
```

**Reply by veewee** [reply]
Based on the spec, example payloads and much more resources, I've crafted a first PSR18 plugin that supports MTOM and SWA for both SOAP requests and responses:

https://github.com/php-soap/psr18-attachments-middleware

Let's continue and improve the package from there on.
